### PR TITLE
Add a handheld cell charger for engineers

### PR DESCRIPTION
**Pull request by lbnesquik**
## About The Pull Request
This little device is based on an ancient technology known as a dynamo. And the code of the plasma cutter.
Activate it to add a tiny, tiny amount of power to a given power cell.
I even did sprites for it.

youtube-video-gif

## Why It's Good For The Game
A little gadget for the engi in need of JUST that little bit of power. It feels silly to me that the three options are carrying cells, carrying a generator or fixing the entire colony power supply.

## Changelog
:cl:
add: Added a handheld cell charger. It isn't going to fix all of your needs, but it might just save you in a pinch.
/:cl:

### CODE REVIEW

**Comment by EvilMelonTheSecond** [comment]
The corpsman backpack has a big internal battery doesn't it? An engineer variant for it would've been my first thought but a jumpstart tool is real cool.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 1]
```diff
@@ -463,3 +463,64 @@
...
+	if(!cell)
+		to_chat(src, "You need some cell to be useful, idiot")
+		return
+	if(!do_after(user, 5, TRUE, src, BUSY_ICON_CLOCK))
+		return
+
+		to_chat(src, "You squeeze the handle a few time, putting in a few volts of charge.")
+		cell.charge += 100
+		playsound(user, 'sound/weapons/guns/interact/rifle_reload.ogg', 20, 1, 5)
+		return
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:493]: from l 490 to 493, bad indentation.

Comment (BraveMole) [reply code/game/objects/items/tools/maintenance_tools.dm:493]: return is useless

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 2]
```diff
@@ -463,3 +463,64 @@
...
+	flags_atom = CONDUCT
+	force = 6.0
+	throw_speed = 2
+	throw_range = 9
+
+	flags_equip_slot = ITEM_SLOT_BELT
+	materials = list(/datum/material/metal = 50, /datum/material/glass = 20)
+	var/obj/item/cell/cell
+
+/obj/item/tool/handheld_charger/attack_self(mob/user as mob)
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:483]:
```suggestion
/obj/item/tool/handheld_charger/attack_self(mob/user)
```

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 3]
```diff
@@ -463,3 +463,64 @@
...
+	item_state = "handheldcharger"
+	w_class = WEIGHT_CLASS_SMALL
+	flags_atom = CONDUCT
+	force = 6.0
+	throw_speed = 2
+	throw_range = 9
+
+	flags_equip_slot = ITEM_SLOT_BELT
+	materials = list(/datum/material/metal = 50, /datum/material/glass = 20)
+	var/obj/item/cell/cell
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:481]: autodoc

Comment (BraveMole) [reply code/game/objects/items/tools/maintenance_tools.dm:481]: and cell needs to be qdelled when the tool is destroyed

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 4]
```diff
@@ -463,3 +463,64 @@
...
+		var/replace_install = "You replace the cell in [src]"
+		if(!cell)
+			replace_install = "You install a cell in [src]"
+		else
+			cell.update_icon()
+			user.put_in_hands(cell)
+		cell = I
+		to_chat(user, span_notice("[replace_install] <b>Charge Remaining: [cell.charge]/[cell.maxcharge]</b>"))
+		playsound(user, 'sound/weapons/guns/interact/rifle_reload.ogg', 20, 1, 5)
+		icon_state = "handheldcharger"
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:512]: this should probably be in update_icon_state, and here you should call update_icon()

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 4 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:512]: We settled on leaving it as it is on discord

**Comment by lbnesquik** [comment]
Suggestions implemented

**Comment by Losenis** [comment]
Give one to the synth as well in his essential kit, given he's also meant to be able to fulfill a non-combat engi role.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 5]
```diff
@@ -463,3 +463,67 @@
...
+		return
+
+	cell.charge += 100
+	to_chat(src, "You squeeze the handle a few time, putting in a few volts of charge.")
+	playsound(user, 'sound/weapons/guns/interact/rifle_reload.ogg', 20, 1, 5)
+
+/obj/item/tool/handheld_charger/attackby(obj/item/I, mob/user, params)
+	. = ..()
+
+	if(istype(I, /obj/item/cell))
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:497]: early return here

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Review by Kuro020 — CHANGES_REQUESTED** [review]
The pallete and shading needs a bit of work

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 6]
```diff
@@ -463,3 +463,68 @@
...
+	materials = list(/datum/material/metal = 50, /datum/material/glass = 20)
+	/// This is the cell we ar charging
+	var/obj/item/cell/cell
+
+/obj/item/tool/handheld_charger/attack_self(mob/user)
+	if(!cell)
+		to_chat(src, "You need some cell to be useful, idiot")
+		return
+	if(!do_after(user, 5, TRUE, src, BUSY_ICON_CLOCK))
+		return
```
Comment (TiviPlus) [inline code/game/objects/items/tools/maintenance_tools.dm:488]: 0.5 SECONDS

Also you can make this loop with a while loop if you dont want people to spam attack self

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 7]
```diff
@@ -463,3 +463,68 @@
...
+	user.put_in_active_hand(cell)
+	cell = null
+	playsound(user, 'sound/machines/click.ogg', 20, 1, 5)
+	to_chat(user, span_notice("You remove the cell from [src]."))
+	icon_state = "handheldcharger_empty"
+	return
+
+/obj/item/tool/handheld_charger/Destroy()
+	qdel(cell)
+	return ..()
```
Comment (TiviPlus) [inline code/game/objects/items/tools/maintenance_tools.dm:530]: QDEL_NULL

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 6 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm]: Implemented

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 7 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm]: Implemented

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 8]
```diff
@@ -463,3 +463,79 @@
...
+	cell = null
+
+/obj/item/tool/handheld_charger/attack_self(mob/user)
+	if(!cell)
+		to_chat(src, "You need some cell to be useful, idiot")
+		return
+
+	if(!recharging)
+		recharging = 1
+		while(recharging)
```
[the remark this answers is not in the document]
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:495]: Implemented

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 9]
```diff
@@ -463,3 +463,79 @@
...
+			if(!do_after(user, 1 SECONDS, TRUE, src, BUSY_ICON_GENERIC))
+				return
+			cell.charge += 200
+			to_chat(user, "You squeeze the handle a few times, putting in a few volts of charge.")
```
[the remark this answers is not in the document]
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:499]: Implemented

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 10]
```diff
@@ -463,3 +463,81 @@
...
+	if(!cell)
+		to_chat(src, "You need some cell to be useful, idiot")
+		return
+
+	if(recharging)///Already using it.
+		recharging = FALSE
+		to_chat(user, "You stop using the recharger.")
+		return
+
+	else
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:498]: else is useless, you did a return

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 11]
```diff
@@ -463,3 +463,81 @@
...
+	if(!cell)
+		replace_install = "You install a cell in \the [src]"
+	else
+		cell.update_icon()
+		user.put_in_hands(cell)
+	cell = I
+	to_chat(user, span_notice("[replace_install]. <b>Charge Remaining: [cell.charge]/[cell.maxcharge]</b>"))
+	playsound(user, 'sound/weapons/guns/interact/rifle_reload.ogg', 20, 1, 5)
+	icon_state = "handheldcharger"
+	return
```
Comment (BraveMole) [inline code/game/objects/items/tools/maintenance_tools.dm:526]: useless return

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 10 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm]: Outdated.

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 11 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:532]: Fixed. Thank you.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 12]
```diff
@@ -463,3 +463,86 @@
...
+		recharging = FALSE
+		to_chat(user, span_notice("You stop using the recharger."))
+		return
+
+	if(cell.charge >= cell.maxcharge)
+		to_chat(user, span_notice("\The [cell] is already fully charged."))
+		return
+	recharging = TRUE
+	while(do_after(user, 1 SECONDS, TRUE, src, BUSY_ICON_GENERIC))
+		cell.charge += 200
```
Comment (TiviPlus) [inline code/game/objects/items/tools/maintenance_tools.dm:503]: Make sure this cant go over cell.maxcharge by using a min() check

cell.charge = min(cell.charge + 200, cell.maxcharge)

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 13]
```diff
@@ -463,3 +463,86 @@
...
+	if(!cell)
+		return ..()
+	cell.update_icon()
+	user.put_in_active_hand(cell)
+	cell = null
+	playsound(user, 'sound/machines/click.ogg', 20, 1, 5)
+	to_chat(user, span_notice("You remove the cell from [src]."))
+	icon_state = "handheldcharger_empty"
+	return
+
```
Comment (TiviPlus) [inline code/game/objects/items/tools/maintenance_tools.dm:545]:
```suggestion
```

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 12 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:503]: Implemented. Thanks

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 13 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:545]: Done.

**Comment by lbnesquik** [comment]
> 
> 
> Looks fine otherwise just get kuro to approve the sprite

Kuro took a look and said they like the black one

**Review by Kuro020 — APPROVED** [review]

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 14]
```diff
@@ -463,3 +463,86 @@
...
+	w_class = WEIGHT_CLASS_SMALL
+	flags_atom = CONDUCT
+	force = 6.0
+	throw_speed = 2
+	throw_range = 9
+	flags_equip_slot = ITEM_SLOT_BELT
+	materials = list(/datum/material/metal = 50, /datum/material/glass = 20)
+	/// This is the cell we ar charging
+	var/obj/item/cell/cell
+	var/recharging = FALSE
```
Comment (TiviPlus) [inline code/game/objects/items/tools/maintenance_tools.dm:482]: Autodoc

**Review by TiviPlus — APPROVED** [review]

File: code/game/objects/items/tools/maintenance_tools.dm (hunk 14 above)
Comment (lbnesquik) [reply code/game/objects/items/tools/maintenance_tools.dm:483]: Implemented

**Pull request merged**